nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py

```diff
@@ -193,9 +201,36 @@ def loss_func(self, pi_logprobs, ref_logprobs, labels, average_log_probs=False):
         rewards = self.get_reduced_masked_logps(
             pi_logprobs - ref_logprobs, labels, average_log_probs=average_log_probs
         )
-        chosen_rewards, reject_rewards = self.split_output_tensor(self.ref_policy_kl_penalty * rewards)
 
-        loss = -torch.nn.functional.logsigmoid(chosen_rewards - reject_rewards)
+        chosen_rewards, reject_rewards = self.split_output_tensor(rewards)
+        logits = chosen_rewards - reject_rewards
+        assert self.preference_loss in ["dpo", "ipo", "kto"]
```

Comment:
can you put the implementation of these 3 loss function in separate functions, and then put these functions in a dictionary?

something like:


```python
PREFERENCE_LOSS_FUNCTIONS = {
"dpo": dpo_loss_function,
...
}
```
and then get it in the model `__init__`?

Reply:
I put this dict outside the class. Alternatively, we can also make it a part of the MegatronGPTDPOModel class.